### ISSUE TITLE
Strange equality check for PostgreSQL arrays with unicode string characters

### ISSUE THREAD

**Issue by smolnar** [rails/rails#22730]
Recently I stumbled upon a problem with `changed?` method on array columns in AR. Assigning identical arrays to created AR model generates changes even though arrays are equal. I managed to observe the behaviour only with array containing unicode strings.

More descriptive example with failing test:
```ruby
 def test_change_of_unicode_string_in_array
    x = PgArray.create!(tags: ['nový']) # 'new'

    assert_not(x.changed?)

    x.tags = ['nový']

    assert_not(x.changed?) # Failure: Expected true to be nil or false

    x.reload
    x.tags << 'modrý' # 'blue'

    assert_equal(x.tags, ['nový', 'modrý'])
    assert(x.changed?)

    x.save!

    assert_not(x.changed?)
  end
```

Test passes with value `novy` instead of unicode one `nový`. This probably isn't a proper behaviour. Am I missing some settings or is this a bug?

**Comment by sgrif** [comment]
Can you provide an executable test case using our [standard bug report template](https://github.com/rails/rails/blob/master/guides/bug_report_templates/active_record_master.rb)?

**Comment by smolnar** [comment]
Sure. It looks like that after creating a record, unicode string is returned in encoded format like `["nov\xC3\xBD"]`, which causes the problem, I think. Reloading record after creating it fixes the issue.

```ruby
begin
  require 'bundler/inline'
rescue LoadError => e
  $stderr.puts 'Bundler version 1.10 or later is required. Please update your Bundler'
  raise e
end

gemfile(true) do
  source 'https://rubygems.org'
  gem 'rails', github: 'rails/rails'
  gem 'arel', github: 'rails/arel'
  gem 'rack', github: 'rack/rack'
  gem 'sprockets', github: 'rails/sprockets'
  gem 'sprockets-rails', github: 'rails/sprockets-rails'
  gem 'sass-rails', github: 'rails/sass-rails'
  gem 'pg'
end

require 'active_record'
require 'minitest/autorun'
require 'logger'

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: 'postgresql', database: 'activerecord_unittest') # Default database from rails/activerecord
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true  do |t|
    t.string :tags, array: true
  end

  create_table :comments, force: true  do |t|
    t.integer :post_id
  end
end

class Post < ActiveRecord::Base
  has_many :comments
end

class Comment < ActiveRecord::Base
  belongs_to :post
end

class BugTest < Minitest::Test
  def test_change_of_unicode_string_in_array
    x = Post.create!(tags: ['nový'])

    # x.reload

    assert_equal(false, x.changed?)
    assert_equal(x.tags, ['nový'])

    x.attributes = { tags: ['nový'] }

    assert_equal(false, x.changed?) # Failure: Expected true to be nil or false

    x.reload
    x.tags << 'modrý'

    assert_equal(x.tags, ['nový', 'modrý'])
    assert_equal(true, x.changed?)

    x.save!

    assert_equal(false, x.changed?)
  end
end
```

**Comment by sgrif** [comment]
So this is actually a bug in the Ruby PG gem, not in Rails -- however I'm leaving this open as we're only relying on the PG gem for this in recent versions, and we need to stop doing that if we can't resolve this issue there.

**Comment by sgrif** [comment]
PG issue: https://bitbucket.org/ged/ruby-pg/issues/230/pg-textencoder-array-doesnt-properly

**Comment by smolnar** [comment]
Awesome, thank you for adding the issue!

**Comment by larskanis** [comment]
This is not a third party issue. It's a ActiveRecord issue. There will be no upstream fix. See also [my comment in #23619](https://github.com/rails/rails/pull/23619#issuecomment-189924036).

**Comment by y-yagi** [comment]
Fixed by 7ba3a48df5bfdc5e98506bb829f937e03b55a5b3 ?

**Comment by sgrif** [comment]
Yes
